flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java

```diff
@@ -0,0 +1,407 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.webmonitor.threadinfo;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.executiongraph.AccessExecutionJobVertex;
+import org.apache.flink.runtime.executiongraph.AccessExecutionVertex;
+import org.apache.flink.runtime.executiongraph.ExecutionJobVertex;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerGateway;
+import org.apache.flink.runtime.taskexecutor.TaskExecutorGateway;
+import org.apache.flink.runtime.taskmanager.TaskManagerLocation;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.runtime.webmonitor.stats.OperatorStatsTracker;
+import org.apache.flink.runtime.webmonitor.stats.Statistics;
+
+import org.apache.flink.shaded.guava18.com.google.common.cache.Cache;
+import org.apache.flink.shaded.guava18.com.google.common.cache.CacheBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Tracker of thread infos for {@link ExecutionJobVertex}.
+ *
+ * @param <T> Type of the derived statistics to return.
+ */
+public class ThreadInfoOperatorTracker<T extends Statistics> implements OperatorStatsTracker<T> {
+
+    /**
+     * Create a new {@link Builder}.
+     *
+     * @param createStatsFn Function that converts a thread info sample into a derived statistic.
+     *     Could be an identity function.
+     * @param <T> Type of the derived statistics to return.
+     * @return Builder.
+     */
+    public static <T extends Statistics> Builder<T> newBuilder(
+            GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
+            Function<OperatorThreadInfoStats, T> createStatsFn,
+            ExecutorService executor) {
+        return new Builder<>(resourceManagerGatewayRetriever, createStatsFn, executor);
+    }
+
+    /**
+     * Builder for {@link ThreadInfoOperatorTracker}.
+     *
+     * @param <T> Type of the derived statistics to return.
+     */
+    public static class Builder<T extends Statistics> {
+
+        private final GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever;
+        private final Function<OperatorThreadInfoStats, T> createStatsFn;
```

Comment:
unless I'm missing something this is always `Function.Identity()`.

Reply:
Well, for the case of assembling the Flame Graph yes, but the idea was to be able to pass in whatever aggregation that might be needed (averaging/median etc.) for other individual parameters that could be contained in the returned samples.

Reply:
I left as is.